### DISCUSSION TITLE
Unit testing a function from conftest.py

### DISCUSSION THREAD

**Discussion by jandom** [pytest-dev/pytest#8915] [Q&A] [answered]
This is a very basic question but I wasn't able to find a relevant answer anywhere here

My conftest.py has the helper inside, to auto-apply some markers 

```
def pytest_collection_modifyitems(items):
    for item in items:
        if 'foo' in item.parent.name:
            item.add_marker(pytest.mark.require_foo)
        if 'bar' in item.parent.name:
            item.add_marker(pytest.mark.require_bar)
```

I would love to factor this method to live in a `utilities.py` but I have no idea how to unit test this? The simplest thing would be to in-place create a list `[pytest.Item, pytest.Item]` which are some hand-crafted test objects. However, that's not super-easy to do, I get the following error message 

```
Direct construction of Item has been deprecated, please use Item.from_parent.
```

Any tips on how to unit test this code? I could created baroque mocks but would prefer not to...

**Comment by nicoddemus** [comment, marked as answer]
Yeah `Item` is not meant to be instantiated directly.

One aspect that is often overlooked is that you don't need to test actual Item objects, but any object which implements the same interface/protocol.

You can use [`mock.create_autospec`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.create_autospec) to create a mock object with the same interface.

**Reply by jandom** [reply]
Thanks @nicoddemus this was very helpful! Posting a useful snippet for anybody else wanting to test their pytest utility code 

```python 
from conftest_utilities import collection_modifyitems
import pytest
from unittest import mock

def test_collection_modifyitems_invalid():
    mock_parent = mock.create_autospec(pytest.Item)
    mock_parent.name = 'foo'
    mock_item = mock.create_autospec(pytest.Item)
    mock_item.parent = mock_parent
    collection_modifyitems([mock_item])
    assert mock_item.add_marker.call_args[0] == (pytest.mark.require_foo,)
```
